c/meterpreter/source/extensions/stdapi/server/net/socket/tcp_server.c

```diff
@@ -19,6 +19,52 @@
 extern IN6_ADDR in6addr_any;
 #endif
 
+static int get_ai_family(const char* address) {
+	struct addrinfo* resolved_host = NULL;
+	struct addrinfo hints = {
+		.ai_family = AF_UNSPEC,
+		.ai_flags = AI_NUMERICHOST
+	};
+
+	int ai_family = AF_UNSPEC;
+	if (getaddrinfo(address, NULL, &hints, &resolved_host) == 0) {
+		ai_family = resolved_host->ai_family;
+		freeaddrinfo(resolved_host);
+	}
+	return ai_family;
+}
+
+// The inet_pton function is not available prior to Windows 8.1 so
+// use getaddrinfo.
+static int inet_pton(int af, const char* src, void* dst) {
+	struct addrinfo* resolved_host = NULL;
+	struct addrinfo hints = {
+		.ai_family = af,
+		.ai_flags = AI_NUMERICHOST
+	};
+
+	if ((af != AF_INET) && (af != AF_INET6)) {
+		return -1;
+	}
+	
+	int ai_family = AF_UNSPEC;
+	if (getaddrinfo(src, NULL, &hints, &resolved_host)) {
+		return 0;
+	}
+	if (resolved_host->ai_family != af) {  // verify the returned address is of the expected family
+		freeaddrinfo(resolved_host);
+		return 0;
+	}
+	if (af == AF_INET) {
+		memcpy(dst, &((struct sockaddr_in*)resolved_host->ai_addr)->sin_addr, sizeof(struct in_addr));
+	} else if (af == AF_INET6) {
+		memcpy(dst, &((struct sockaddr_in6*)resolved_host->ai_addr)->sin6_addr, sizeof(struct in_addr6));

```

Comment:
Would be ideal to convert these `memcpy` calls to `memcpy_s` calls for safety.

Reply:
On second observation we also aren't checking if these memcpy's complete successfully prior to proceeding with the code which is something else we should be checking for here.

Reply:
> Would be ideal to convert these memcpy calls to memcpy_s calls for safety.

We don't have a size for `dst` since it's a `void *` so it'd be `memcpy_s(dst, sizeof(struct in_addr), &((struct sockaddr_in*)resolved_host->ai_addr)->sin_addr, sizeof(struct in_addr))` which doesn't seem any more secure. Let me know if you find a code path where `sizeof(dst) < min(sizeof(struct in_addr), sizeof(struct in_addr6))`.

> On second observation we also aren't checking if these memcpy's complete successfully prior to proceeding with the code which is something else we should be checking for here.

Sure I can add that check in. I guess it could come up in the future if the function is reused.

Reply:
Thanks and yeah agree on your `memcpy_s` point, overlooked that a bit. My thought was to see if we could get the size of `dst` itself but this would require us updating the `inet_pton` function signature to take in a size parameter which would likely cause incompatibility issues.